Share torrents until seeding time reaches an specific amount of minutes

naikel opened the pull request:
This feature appears from time to time in the Issues section as a [Wishlist].  I use it heavily.

This allows to specify seeding a torrent for up to X minutes.  You can also mix it with the max share ratio option and the torrent will stop seeding if any of both rules applies first.

This PR has several changes:

1.- The method processBigRatios() was renamed to processShareLimits().  That's because the name was misleading in the first place since most people set "1" or "2" as a share ratio limit and those are not big numbers, and also because this method now also process seeding time limits.

2.- Sadly I couldn't change the name of max_ratio_act for the action to apply to the torrent when it's completed in the WebUI because that would break backward compatibility.

3.- The ETA method now returns qMin(ratio_eta,seeding_time_eta), that is the lowest time between reaching the ratio or reaching the time limit.

4.- The option dialog was changed to this:

![qBittorrent Options](http://i.imgur.com/IoiKlgj.png)

5.- The WebUI was changed to this:

![WebUI Options](http://i.imgur.com/pEFPhi3.png)

6.- And finally the torrent share settings to:

![WebUI Torrent Setting](http://i.imgur.com/u0yPXaH.png)

It works great!  I've been using it for months.

Hope you like it.

naikel commented:
Reverted the unnecessary alignment changes of options.ui.

glassez commented on src/core/bittorrent/torrenthandle.cpp:
```diff
@@ -828,16 +838,29 @@ qulonglong TorrentHandle::eta() const
-
         qreal max_ratio = maxRatio();
-        if (max_ratio < 0) return MAX_ETA;
+        int max_seeding_time = maxSeedingTime();
```
Use camelCase in new code. See CODING_GUIDELINES.md.

naikel replied:
Ok, will do, but only half of the original code is camelCased.  Just the C++ source code right? JavaScript will be a mess if it's camelCased.

glassez replied:
>but only half of the original code is camelCased

Rules have been developed recently. We use them in new code and changing old from time to time.

sledgehammer999 commented:
Rebase.

naikel commented:
Rebased and fixed camelCase.

sledgehammer999 commented:
@glassez I think we should relax the camelCase rule. It is a pain in the ass when you are moving old code around without actually touching it but you have to read it and fix it to be camelCased.

glassez commented:
>It is a pain in the ass when you are moving old code around without actually touching it but you have to read it and fix it to be camelCased.

Otherwise it will be a pain in the ass for reviewers to distinguish new code from displaced. I have no trouble in correcting such names. QtCreator does it perfectly. Just have to find the wrong names, but it is also not difficult: they just cut look. Besides, if someone misses something, the other reviewers remind him.

Nemo-qB commented:
Looks great naikel! Very useful when seeding.

naikel commented:
@Nemo-qB Thanks!

I rebased and made a little cosmetic change since the word "minutes" was missing in the WebUI:

![WebUI Preferences](http://i.imgur.com/n8h1ODW.png)

naikel commented:
Fixed this line to have a default value:

```C++
out.seedingTimeLimit = fast.dict_find_int_value("qBt-seedingTimeLimit", -2);
```

Because without it, people who upgraded from 3.3.1 to this version and had seeding torrents would see them all stopped because they would have reached the max share time ratio of 0 minutes!

But now upgrades work great as well.

naikel commented:
Rebased.

No love for this PR? :cry:

naikel commented:
@sledgehammer999 I wonder if this can make it to 3.3.2 or 3.4.0?

LordNyriox commented:
@naikel Personally, I've been wanting to see this feature since forever!  Maybe nobody else realizes this, but when you're on a tight schedule and seeding ratios are not a guarantee within said schedule, a feature like this would be a lifesaver.

naikel commented:
Rebased.

@LordNyriox you can clone this branch in my repository and use the feature until @sledgehammer999 merges it.  I've been using it since October, can't live without it!

LordNyriox commented:
Only if I could figure out how to build the code in the first place.
Unfortunately, no guide in the world will make sense to me unless someone
(quite literally) holds my hands through the entire process the first time.


On Tuesday, January 12, 2016, Scorpius <notifications@github.com> wrote:

> Rebased.
>
> @LordNyriox <https://github.com/LordNyriox> you can clone this branch in
> my repository and use the feature until @sledgehammer999
> <https://github.com/sledgehammer999> merges it. I've been using it since
> October, can't live without it!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qbittorrent/qBittorrent/pull/4266#issuecomment-171073301>
> .
>

LordNyriox commented:
@naikel Thanks for the advice. Unfortunately for me, I'm a klutz with code, no matter what I try to do with it (including compile).  So right now I'm at the tender mercy of Gelmir to compile builds.  (He builds 64-bit Windows binaries, which are a lifesaver.)  

Anyway, Github is whining about branch conflicts again.  Do you think you could rebase?  

Thanks, 
~ Nyriox

naikel commented:
Rebased and retested.

@sledgehammer999 you should merge this when you have the time.

LordNyriox commented:
@naikel One request:  Do you think you can squash all your commits into one for this PR?  I think it was called 'git fixup' or something like that.  

Thanks again for the wonderful work! 
~ Nyriox

LordNyriox commented:
@naikel Thanks.

LordNyriox commented:
Ugh.  @naikel, I hate to break this to you, but Github's whining about metge conflicts again.  Can you rebase again, please? 

Thanks, 
~ Nyriox

naikel commented:
Sure, but will this ever be merged? It's ready since Oct 27th...

LordNyriox commented:
@naikel There are PRs just as old that have only now been merged.  Right now @sledgehammer999 is busy with [https://github.com/qbittorrent/qBittorrent/pull/4681], trying to make the logging feature more useful.  Once that's done he's still got dozens of other PRs to review, and not enough spare time to review them all in.  

So, to answer your question, I see no reason for it not to be merged eventually, so long as it's kept up to date with current code.  There are enough people who can use this feature for it to remain completely ignored.  

Thanks, 
~ Nyriox

naikel commented:
I can't rebase because "master" right now crashes when I click "Options", so it's totally unstable :cry:

LordNyriox commented:
@naikel That sucks… But do you really think it might be better to wait until later to try rebasing again?  There have been a lot of changes to the Options layout over the last couple of weeks, including a teardown of the Advanced section.  

I suspect with all the Options reconfiguring going on, you might be experiencing some kind of positioning conflict or something when you open the window.  You might have to do a thorough review of what's changed in the options window recently before you can track down the problem.  Of course, I expect that'll take some time.  

Thanks for all your great work! 

~ Nyriox

naikel commented:
@LordNyriox what I meant is that the master branch here on github without any of my patches crashes when I click "Options".
```
ASSERT: "ok" in file gui/advancedsettings.cpp, line 324


*************************************************************
Catching signal: SIGABRT
Please file a bug report at http://bug.qbittorrent.org and provide the following information:

qBittorrent version: v3.4.0alpha
stack trace:
  /lib/x86_64-linux-gnu/libc.so.6 : gsignal()+0x37  [0x7fcd53b89107]
  /lib/x86_64-linux-gnu/libc.so.6 : abort()+0x148  [0x7fcd53b8a4e8]
  /usr/lib/x86_64-linux-gnu/libQt5Core.so.5 : ()+0x931e1  [0x7fcd54cd71e1]
  /usr/lib/x86_64-linux-gnu/libQt5Core.so.5 : ()+0x8e33e  [0x7fcd54cd233e]
  src/qbittorrent : void AdvancedSettings::addRow<QLabel>(int, QString const&, QLabel*)+0x69  [0x6540a9]
  src/qbittorrent : AdvancedSettings::loadAdvancedSettings()+0x82  [0x651e72]
  src/qbittorrent : AdvancedSettings::AdvancedSettings(QWidget*)+0x3a7  [0x653dd7]
  src/qbittorrent : options_imp::options_imp(QWidget*)+0x2dc7  [0x661787]
  src/qbittorrent : MainWindow::on_actionOptions_triggered()+0x5b  [0x6069bb]
```
So I guess @sledgehammer999 should fix this first until the master branch is stable again.

LordNyriox commented:
@naikel Ouch.  One of his most recently merged PRs was a complete overhaul of the Advanced Settings code.  You might want to ask him what might be causing an issue on your system.  (GNU Linux, if I'm reading the crash right.)  

@sledgehammer999 @glassez You guys might want to verify that @Chocobo1's Advanced Settings overhaul is clean, because @naikel is showing a crash log referencing `Advanced Settings` every time he opens up the Options window with current master.  

Thanks, 
~ Nyriox

naikel commented:
I actually see what the problem is.  

The new addRow() method in AdvancedSettings is assigning a boolean as a return vale of connect, but connect doesn't return a boolean, but a QMetaObject::Connection instead.  So the assert at the end will always hit:
```C++
    bool ok;
    if (std::is_same<T, QCheckBox>::value)
        ok = connect(widget, SIGNAL(stateChanged(int)), SIGNAL(settingsChanged()));
    else if (std::is_same<T, QSpinBox>::value)
        ok = connect(widget, SIGNAL(valueChanged(int)), SIGNAL(settingsChanged()));
    else if (std::is_same<T, QComboBox>::value)
        ok = connect(widget, SIGNAL(currentIndexChanged(int)), SIGNAL(settingsChanged()));
    else if (std::is_same<T, QLineEdit>::value)
        ok = connect(widget, SIGNAL(textChanged(QString)), SIGNAL(settingsChanged()));
    Q_ASSERT(ok);
```
This only happens on debug builds.  I will continue my rebase, but this has to be fixed...

Chocobo1 commented:
>This only happens on debug builds. I will continue my rebase, but this has to be fixed...

Sorry about that, please do fix it.

LordNyriox commented:
@naikel FYI a PR by @glassez [https://github.com/qbittorrent/qBittorrent/pull/4772] is now set to fix the issue.  

Thanks, 
~ Nyriox

naikel commented:
Cool!! :+1:

LordNyriox commented:
@naikel:  A whole bunch of PRs were just merged.  Do you think you could rebase this?  

Thanks, 
~ Nyriox

naikel commented:
I'll work on it

LordNyriox commented:
@sledgehammer999:  I know I, personally, have been waiting forever for this particular feature.  Might this get merged some time before 3.3.4?  Thanks.

ngosang commented:
Tested and I can't find anything wrong. @sledgehammer999 could you tag for 3.3.5?

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -1800,21 +1851,22 @@ bool Session::isKnownTorrent(const InfoHash &hash) const
-    if ((m_globalMaxRatio == -1) && !hasPerTorrentRatioLimit()) {
-        if (m_bigRatioTimer->isActive())
-            m_bigRatioTimer->stop();
+    if (((m_globalMaxRatio == -1) && !hasPerTorrentRatioLimit()) &&
```
Fix coding style. Move `&&` operator to next line.

glassez commented on src/gui/updownratiodlg.cpp:
```diff
@@ -28,45 +28,95 @@
     handleRatioTypeChanged();
 }
 
+void UpDownRatioDlg::accept() {
```
Fix coding style in this method.

naikel commented:
Coding style fixed where noted.

glassez commented:
>Coding style fixed where noted.
Still not fixed in `void UpDownRatioDlg::accept()` method body.

naikel commented:
> Still not fixed in void UpDownRatioDlg::accept() method body.

I put the { in the next line like you told me.  Is there anything else left to fix there?

glassez commented:
>EDIT: I think I got it: the if.

Yes. But it still not fixed.
Please check out CODING_GUIDELINES.md to begin with, otherwise you'd have a lot of time to "fix" it.

naikel commented:
I see it. Maybe this is the only project you work on, but when you work in dozens of projects, each one with a coding style totally different than the next one, it's pretty hard to remember.

glassez commented:
>I see it. Maybe this is the only project you work on, but when you work in dozens of projects, each one with a coding style totally different than the next one, it's pretty hard to remember.

And what do you suggest?

Still unfixed. Either drop braces in `if` clause or add them to `else`.

naikel commented:
Thanks.  I chose to drop them then.

ngosang commented:
@naikel After this is merged I will love to have a similar feature to stop sharing when one torrent have more than XX seeds. What do you think?

DrKittens commented:
Personally i'd think that would be a seperate PR as its unrelated to this time-based PR.
I'd love to see this functional 5month old PR merged into master branch sometime soon please.

LordNyriox commented:
@naikel:  Some significant changes to the Options UI were merged recently.  Can you rebase this PR, please?  

Thanks.

friendlytherapist commented:
@naikel what happened to development on this?

naikel commented:
I rebased this PR again.

I have to admit I'm a little sad this PR has never been considered for a merge.  It is an important feature.  Not everybody can seed forever.  Some of us have a high rotation of short-lived torrents and this feature works extremely great (with another PR I haven't published where a completed torrent means a torrent that has finished the download AND the seeding).  Some others have limited storage space and have to get rid of their downloads quickly.  Others have an extremely expensive upload rate and can't bear to keep torrents seeding forever.  There's always a request for this feature in the issues section, always, since the beginning of this project.  I'm literally amazed this hasn't been merged.

I have a setting of 15 days (in minutes) and works perfectly.  

Looks like the main developing team is not interested in this PR.  I try to keep it rebased from time to time for the people that is compiling qBittorrent by themselves and merging this PR individually, and I know several people that do this.

LordNyriox commented:
@qbittorrent-frequent-contributors, @onto, @Chocobo1, @evsh, @glassez, @sledgehammer999:  Can someone please do a review on this PR to see how it measures to current code?  

@naikel:  Why not publish the second PR you mentioned as an "Advanced Option" or something like that?  It certainly could be a useful debug feature.  

Thanks again! 
~ Nyriox 

_____

naikel commented:
@LordNyriox The second PR I mentioned changes the definition of the completed status of a torrent and the main developers do not agree with this vision. I'll elaborate:

What I think is natural is this:

- Downloading: A torrent that is currently downloading
- Seeding: A torrent that finished downloading and it is currently seeding
- Completed: A torrent that finished downloading and seeding, so it's paused because its job is done (blue paused icon)
- Resumed: A torrent that is either downloading or seeding, not necessarily active.
- Paused: A torrent that is paused because user paused it or because it's completed.
- Active: A torrent that is currently transferring something.
- Inactive: A torrent that is not transferring anything.

But the qBittorrent defines the completed status like this:

- Completed: A torrent that finished downloading (but might be seeding)

And this is redundant with seeding torrents because a torrent has to have finished the download to be seeding. So that's a mistake in my humble opinion.  A completed torrent must have finished all their jobs, downloading AND seeding.

So I have this nice PR where the Completed torrents section include only torrents that are really finished (download and upload), mostly because they reached their share ratio or the seeding time limit.

naikel commented:
@LordNyriox not really.  If you look at the code it says:

```
bool TorrentHandle::isCompleted() const
{
    return m_state == TorrentState::Uploading
            || m_state == TorrentState::StalledUploading
            || m_state == TorrentState::CheckingUploading
            || m_state == TorrentState::PausedUploading
            || m_state == TorrentState::QueuedUploading
            || m_state == TorrentState::ForcedUploading;
}
```

But it should be:
```
bool TorrentHandle::isCompleted() const
{
    return m_state == TorrentState::PausedUploading;
}
```
And that's exactly what my PR does (that and in the WebUI code as well)

LordNyriox commented:
@naikel:  What I was trying to suggest is a (cosmetic) PR making an OPTION to show the `Completed` status only if a torrent has reached its seeding limit as well as reached 100%.

naikel commented:
@LordNyriox Oh...

Good idea! I have to admit I never thought of that... 

I get it now.  It can go in the Advanced Options section, like you suggested.

thalieht commented:
Only in minutes? We'd need a calculator to determine the days.

naikel commented:
@thalieht don't be lazy.

zywo commented:
Unfortunately, the merging of many useful PRs is very very slow for a project like qBittorrent.

thalieht commented on src/gui/updownratiodlg.cpp:
```diff
@@ -28,49 +28,94 @@
         ui->useDefaultButton->setChecked(true);
-    }
-    else if (initialValue == -1) {
+    } else if (initialRatioValue == -1 && initialTimeValue == -1) {
```
Coding style

thalieht commented on src/gui/updownratiodlg.cpp:
```diff
@@ -28,49 +28,94 @@
-    else {
+        initialRatioValue = BitTorrent::Session::instance()->globalMaxRatio();
+        initialTimeValue = BitTorrent::Session::instance()->globalMaxSeedingMinutes();
+    } else {
```
Coding style

naikel commented:
Fixed coding style.

thalieht commented on src/base/bittorrent/session.cpp:
```diff
@@ -1285,36 +1302,65 @@ void Session::populateAdditionalTrackers()
+                qreal ratioLimit = torrent->ratioLimit();
+                if (ratioLimit == TorrentHandle::USE_GLOBAL_RATIO)
+                    // If Global Max Ratio is really set...
+                    ratioLimit = this->globalMaxRatio();
```
this-> is not necessary.

naikel replied:
True, that's the result of the search/replace from the previous implementation that used a global variable instead.  I'll fix it.

thalieht commented on src/base/bittorrent/session.cpp:
```diff
@@ -1285,36 +1302,65 @@ void Session::populateAdditionalTrackers()
+                int seedingTimeLimit = torrent->seedingTimeLimit();
+                if (seedingTimeLimit == TorrentHandle::USE_GLOBAL_SEEDING_TIME)
+                     // If Global Seeding Time Limit is really set...
+                    seedingTimeLimit = this->globalMaxSeedingMinutes();
```
this-> is not necessary.

naikel replied:
Same, see above.  I'll fix it.

thalieht commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -190,7 +192,11 @@ TorrentState::operator int() const
+const int TorrentHandle::NO_SEEDING_TIME_LIMIT = -1;
+
 const qreal TorrentHandle::MAX_RATIO = 9999.;
+const int TorrentHandle::MAX_SEEDING_TIME = 525600;
```
Why limit it to a year? Sure, probably noone will set it even to a year but it should be the user's choice.

naikel replied:
I have to put a number in there, like the max ratio has one.  I chose the first coherent option I could think of.

thalieht commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -1055,6 +1081,23 @@ qreal TorrentHandle::maxRatio(bool *usesGlobalRatio) const
     return ratioLimit;
 }
 
+int TorrentHandle::maxSeedingTime(bool *usesGlobalSeedingTime) const
```
Why not a reference?

evsh replied:
I would use `std::pair<int, bool> maxSeedingTime()` and then the function would be:
```C++
std::pair<int, bool> TorrentHandle::maxSeedingTime() const
{
    return m_seedingTimeLimit == USE_GLOBAL_SEEDING_TIME ?
        std::make_pair(m_session->globalMaxSeedingMinutes(), true) :
        std::make_pair(m_seedingTimeLimit, false) :
}

naikel replied:
I tried to follow the same coding style that was already there.  Look for maxRatio() and other similar functions that has the same implementation.  Remember this PR was already approved, but never merged.  This is just a rebase actually, but I really appreciate your comments, they have been very helpful.

thalieht replied:
My bad, didn't look at the other methods.

thalieht commented on src/gui/optionsdlg.cpp:
```diff
@@ -1182,6 +1206,9 @@ bool OptionsDialog::useAdditionDialog() const
     applyButton->setEnabled(true);
+
+    // Verify if the share action button must be enabled
+    m_ui->comboRatioLimitAct->setEnabled(m_ui->checkMaxRatio->isChecked() || m_ui->checkMaxSeedingMinutes->isChecked());
```
I doubt this will be accepted here. Maybe it's better to make a new method with it and connect it with the checxkboxes somewhere in the constructor?

```
connect(m_ui->checkMaxRatio, SIGNAL(toggled(bool)), this, SLOT(toggleComboRatioLimitAct()));
connect(m_ui->checkMaxSeedingMinutes, SIGNAL(toggled(bool)), this, SLOT(toggleComboRatioLimitAct()));

void OptionsDialog::toggleComboRatioLimitAct()
{
    m_ui->comboRatioLimitAct->setEnabled(m_ui->checkMaxRatio->isChecked() || m_ui->checkMaxSeedingMinutes->isChecked());
}
```

naikel replied:
It was already accepted before (the whole PR was).  I don't see a problem with this method.

thalieht replied:
Oh ok, i just found it strange to enable/disable that button every time something is changed in the options.

naikel replied:
:) That's because you haven't tried the PR, just read the code.  If neither the option "Seed torrents until their ratio reaches" nor  "Seed torrents until their seeding time reaches" are enabled, the action (where you select if you want to Pause or Remove the torrent) should be disabled.  It's the same functionality than before but now you have to select either one of the options.  

Originally if you select "Seed torrents until their ratio reaches" enables the Action drop-down field and if you unselect it, it disables it.  Same thing but now with two options.

I welcome to try the PR.  It works great and it's a really useful feature :)

thalieht replied:
I did try it and understand what it does otherwise i wouldn't make a comment. I meant that you put it in `enableApplyButton()` which is executed when pretty much anything is changed in the options.

glassez replied:
>That's because you haven't tried the PR, just read the code.

I also have not tested this, but I don't understand why the following code won't work?
```c++
connect(m_ui->checkMaxRatio, SIGNAL(toggled(bool)), this, SLOT(toggleComboRatioLimitAct()));
connect(m_ui->checkMaxSeedingMinutes, SIGNAL(toggled(bool)), this, SLOT(toggleComboRatioLimitAct()));

void OptionsDialog::toggleComboRatioLimitAct()
{
    m_ui->comboRatioLimitAct->setEnabled(m_ui->checkMaxRatio->isChecked() || m_ui->checkMaxSeedingMinutes->isChecked());
}
```
Now you have it in the wrong place. If before no one noticed, that is no reason not to fix it.

glassez replied:
>Because it won't reenable the Apply button if you modify the seeding limit, just the comboRatioLimitAct, and then you wouldn't be able to apply your changes.

But you have connected it to `enableApplyButton()` too, isn't it?

naikel commented:
Removed the unnecessary this-> references.

naikel commented:
It seems Travis builds were fixed so this PR is ready to be merged (again).

LordNyriox commented:
@sledgehammer999: ?

naikel commented:
Still no love for this PR?

RichardWilliamPearse commented:
Just a user comment here: I'd really like to use this functionality too.

sledgehammer999 commented:
@naikel sorry for the delays. I am currently in the process of reviewing+merging a lot of **old** PRs. Eventually I'll get to this one too. I hope you can wait.

naikel commented:
@sledgehammer999 I hope this can make it for 3.3.12!

Nivius commented:
I have been waiting for this function for ages now. and so have loads of friends and colleagues. 

This is the last feature that qbittorrent need to turn from a great client to THE client imo. 

Hope this makes it into 3.3.12!

༼ つ ◕_◕ ༽つ @sledgehammer999  TAKE MY ENERGY ༼ つ ◕_◕ ༽つ

thalieht commented:
Tagged issues to add in a future commit or PR name to close them.
Not sure about the syntax though: "closes #1646 #4519..." or "closes #1646 closes #4519..."
#1646 #4519 #4897 #5815 #6153

evsh commented:
@thalieht thank you, will close all those issues as duplictes.

LordNyriox commented:
@sledgehammer999 approveFeature

Nivius commented:
Is t here anyway that we can use this feature before its officially included??

LordNyriox commented:
@evsh, @Chocobo1, @glassez, @sledgehammer999:  Review?  

@sledgehammer999 approveFeature

thalieht commented on src/gui/updownratiodlg.cpp:
```diff
@@ -43,34 +46,78 @@ UpDownRatioDlg::UpDownRatioDlg(bool useDefault, qreal initialValue,
         ui->useDefaultButton->setChecked(true);
     }
-    else if (initialValue == -1) {
+    else if (initialRatioValue == -1 && initialTimeValue == -1) {
```
Parentheses
`((initialRatioValue == -1) && (initialTimeValue == -1))`

naikel replied:
Why? there's no such thing in the code guidelines.

thalieht replied:
Heh you're right. It's for readability. I've seen it suggested so many times i thought it is for sure in the guidelines. Either way i've seen at least one more and gonna point it out in a while. Feel free to ignore them if noone else says something about it.

Chocobo1 replied:
>It's for readability. 

+1

glassez replied:
+1

thalieht commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -853,16 +865,30 @@ qulonglong TorrentHandle::eta() const
-        if (speed_average.upload == 0) return MAX_ETA;
+        qreal maxRatioValue = maxRatio();
+        int maxSeedingTimeValue = maxSeedingTime();
+        if (maxRatioValue < 0 && maxSeedingTimeValue < 0) return MAX_ETA;
```
Parentheses
((maxRatioValue < 0) && (maxSeedingTimeValue < 0))

thalieht commented on src/base/bittorrent/torrenthandle.cpp:
```diff
@@ -853,16 +865,30 @@ qulonglong TorrentHandle::eta() const
+
+        qlonglong ratioEta = MAX_ETA;
+
+        if (speed_average.upload > 0 && maxRatioValue >= 0) {
```
Parentheses
((speed_average.upload > 0) && (maxRatioValue >= 0))

naikel commented:
I made the suggested changes and rebased.  Retested everything with several torrents.  Everything look good.

evsh commented:
@naikel, sorry, but it has a conflict again. Could you rebase again, please?

naikel commented:
Rebased again @evsh @LordNyriox

naikel commented:
Jesus Christ I think I'm gonna give up.  You guys are never happy.  This request is from December 5th, 2015.  You had a lot of time to review it.  Now it looks everything is bad again and I have done nothing, just apply your changes every time you guys request them.

I think it's better for all if I just close this.  This will never be merged.  Not even if I fix everything, you guys just don't want this feature.

naikel commented on src/base/bittorrent/session.cpp:
```diff
@@ -1277,36 +1294,65 @@ void Session::populateAdditionalTrackers()
+                    if ((seedingTime <= TorrentHandle::MAX_SEEDING_TIME) && (seedingTime >= seedingTimeLimit)) {
+                        Logger* const logger = Logger::instance();
+                        if (m_maxRatioAction == Remove) {
+                            logger->addMessage(tr("'%1' reached the maximum seeding time you set. Removing...").arg(torrent->name()));
```
You should read the whole file not the diff.  This is a copy/paste from the same section that process the ratio limit that's already there.  I just tried to be consistent with the code that was already there.

This is YOUR code not mine:
```
                        if (m_maxRatioAction == Remove) {
                            logger->addMessage(tr("'%1' reached the maximum ratio you set. Removing...").arg(torrent->name()));
                            deleteTorrent(torrent->hash());
                        }
```
I just tried to be consistent in what was already there.  **DO NOT READ JUST DIFFS READ THE WHOLE FILE**

naikel commented on src/base/bittorrent/session.cpp:
```diff
@@ -1277,36 +1294,65 @@ void Session::populateAdditionalTrackers()
+                            logger->addMessage(tr("'%1' reached the maximum seeding time you set. Removing...").arg(torrent->name()));
+                            deleteTorrent(torrent->hash());
+                        }
+                        else {
```
Read the whole file please, the original else is like that.  I can fix MY else though but it would be consistent with the rest of the file.

naikel commented:
@glassez 

> I also have not tested this, but I don't understand why the following code won't work?

Because it won't reenable the Apply button if you modify the seeding limit, just the comboRatioLimitAct, and then you wouldn't be able to apply your changes.

glassez commented:
@naikel, calm down, please! 
In fact, this PR is close to being merged like never (this is due to the recent changes in the organization of the project). 
I really had not considered this PR previously due to the fact that I do not could affect the possibility of its merging, and the feature itself is not interesting to me. Now I'm one of those members who can merge PRs collectively without the participation of the package maintainer (who is very busy and has little time for this).

naikel commented:
I'll make the changes.  

But I will also fix the same functions that have the same problem just to be consistent.

What I mean is, if you want me to fix this:
```
    int globalMaxSeedingMinutes = session->globalMaxSeedingMinutes();
    if (globalMaxSeedingMinutes >= 0) {
        // Enable
        m_ui->checkMaxSeedingMinutes->setChecked(true);
        m_ui->spinMaxSeedingMinutes->setEnabled(true);
        m_ui->spinMaxSeedingMinutes->setValue(globalMaxSeedingMinutes);
    }
```
The variable is totally useless, I know.  But why I did it? because just a couple of lines before there's this code:
```
    float globalMaxRatio = session->globalMaxRatio();
    if (globalMaxRatio >= 0.) {
        // Enable
        m_ui->checkMaxRatio->setChecked(true);
        m_ui->spinMaxRatio->setEnabled(true);
        m_ui->comboRatioLimitAct->setEnabled(true);
        m_ui->spinMaxRatio->setValue(globalMaxRatio);
    }
```
Then I have to fix that as well, and that's not my code, but it would look totally weird that both code sections look different.

The same for the "else", I follow the code guidelines in new methods that I introduce, but when I'm modifying an existing one, I could be a little flexible just to follow the code guideline of that method specifically, because I think some other developer is working in fixing all the code in parallel.

So I'll take the liberty to fix the original functions that I copied to be consistent

LordNyriox commented:
@naikel:  

> Jesus Christ I think I'm gonna give up. You guys are never happy.  

Actually, @glassez and @evsh are hard on everyone when it comes to code quality.  They do not consider code ready until the author hammers out even the slightest possible inconsistency…  

Trust me, I have watched the discussion on every PR in this project for almost two years.  This bunch is tough on coders, but your code will thank you.

naikel commented:
@LordNyriox I understand, and they should be strict since I'm strict with the code at my job as well, but even so,  something like 75% of the code doesn't follow the guidelines...  I wonder what happened.

LordNyriox commented:
@naikel:  

> But I will also fix the same functions that have the same problem

Excellent plan!  In fact, the main devs do this all the time.  

Most of the time, hammering out inconsistencies and fixing bugs in this project are handled as bonus features in PRs that are only vaguely related to the fixes anyway…

LordNyriox commented:
@naikel:  

> something like 75% of the code doesn't follow the guidelines... 

The coding guidelines are actually a relatively recent addition relative to the history of the project.  

I am not certain exactly when they were implemented, but I think it was after `@sledgehammer999` took over from `@cdumez`.  (Most of the base code was written long before that.)  

Also, the coding guidelines themselves have evolved in the recent past, especially after qBittorrent changed to C++11.

Nivius commented:
@naikel you can do it! this is a highly requested feature for a lot of people, it's just that most people don't go this deep to ask for it

This addition whould make my my process 100% automate from 90% right now. I really hope that you cna find strength to work it thru!

@LordNyriox Is it @naikel responsibility to fix code that is not relevant to his part and have that block the merging for his part?

glassez commented:
@LordNyriox, thank you!
@naikel, really the tightening of coding style has been done recently and there is still a lot of old code that doesn't follow it. This is not a reason to break it in new code. Other errors in the code (previously undetected) is also not a reason to emulate them. And Yes, we do try to correct the noticed deficiencies in the code affected in some PR, even if they are unrelated (although requests to do so are not popular with many, including ourselves).

glassez commented:
>I'll make the changes.

@naikel, please create fixup commit first time. You squash it after checking.

evsh commented on src/base/bittorrent/session.cpp:
```diff
@@ -1277,36 +1294,65 @@ void Session::populateAdditionalTrackers()
+
+                if (ratioLimit >= 0) {
+                    qDebug("Ratio: %f (limit: %f)", ratio, ratioLimit);
+                    Q_ASSERT(ratioLimit >= 0.f);
```
Always true, you checked this two lines above.

evsh commented on src/base/bittorrent/session.cpp:
```diff
@@ -1277,36 +1294,65 @@ void Session::populateAdditionalTrackers()
+            }
+
+            if (torrent->seedingTimeLimit() != TorrentHandle::NO_SEEDING_TIME_LIMIT) {
+                const int seedingTime = torrent->seedingTime() / 60;
```
Identifier name in this line is confusing. Suggest
```
const int seedingTimeInMinutes = torrent->seedingTime() / 60;
```

evsh commented on src/gui/optionsdlg.cpp:
```diff
@@ -974,19 +977,32 @@ void OptionsDialog::loadOptions()
     m_ui->checkIgnoreSlowTorrentsForQueueing->setChecked(session->ignoreSlowTorrentsForQueueing());
 
-    if (session->globalMaxRatio() >= 0.) {
+    float globalMaxRatio = session->globalMaxRatio();
```
Negligible part of people use `-flto` with GCC and as such getter has to be defined in the header class to be inlined. But this is prohibited by the coding guidelines in this project.

evsh commented on src/webui/extra_translations.h:
```diff
@@ -74,6 +74,9 @@ static const char *__TRANSLATIONS__[] = {
     QT_TRANSLATE_NOOP("HttpServer", "Unknown"),
     QT_TRANSLATE_NOOP("HttpServer", "Hard Disk"),
     QT_TRANSLATE_NOOP("HttpServer", "Share ratio limit must be between 0 and 9998.")
+    QT_TRANSLATE_NOOP("HttpServer", "Watch Folder"),
```
Why these changes? Maybe a rebase artifact?

naikel commented:
I pushed a new commit with the fixes @evsh @glassez @LordNyriox

LordNyriox commented on src/gui/updownratiodlg.cpp:
```diff
@@ -28,13 +28,16 @@
  * Contact : chris@qbittorrent.org
```
@naikel:  Mostly irrelevant to this PR, but I am pretty sure you should remove this line and edit Line 3 of this file to read:  
`  * Copyright (C) 2011  Christian Kandeler, Christophe Dumez <chris@qbittorrent.org>`.  

Newer code uses that format, and I have seen @glassez change existing files to it in other PRs.

naikel replied:
Not sure to do that change:
```
$ grep -r Kandeler
src/gui/updownratiodlg.cpp: * Copyright (C) 2011  Christian Kandeler, Christophe Dumez
src/gui/updownratiodlg.h: * Copyright (C) 2011  Christian Kandeler, Christophe Dumez
src/base/scanfoldersmodel.h: * Copyright (C) 2010  Christian Kandeler, Christophe Dumez
src/base/scanfoldersmodel.cpp: * Copyright (C) 2010  Christian Kandeler, Christophe Dumez
```
They all are like that

glassez replied:
@LordNyriox, there should be one name per line.
@naikel, leave it as is unless you want to add yourself.

naikel replied:
I left it as it is because I wasn't sure what to do.

glassez commented on src/gui/optionsdlg.cpp:
```diff
@@ -1206,6 +1230,11 @@ void OptionsDialog::enableApplyButton()
     applyButton->setEnabled(true);
 }
 
+void OptionsDialog::toggleComboRatioLimitAct() {
```
Fix curly brace position.

thalieht commented on src/gui/updownratiodlg.cpp:
```diff
@@ -28,13 +28,16 @@
 
+#include <QMessageBox>
+
 #include "updownratiodlg.h"
```
I wasn't gonna say anything because apparently you hate doing coding style but while you're at it reverse these: 
```
#include "updownratiodlg.h"

#include <QMessageBox>
```

Chocobo1 replied:
+1, please fix it too.

evsh commented:
I have no more critical comments. Taking into account the long history of this PR and high user demand, IMO all polishing can be done after merging.

glassez commented:
> have no more critical comments. Taking into account the long history of this PR and high user demand, IMO all polishing can be done after merging.

+1
@naikel, please fix last issue (curly brace position) and squash fixup commits.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -2982,6 +3020,14 @@ bool Session::hasPerTorrentRatioLimit() const
+{
+    foreach (TorrentHandle *const torrent, m_torrents)
+        if (torrent->seedingTimeLimit() >= 0) return true;
+
```
redundant empty line

and sorry for late review...

glassez replied:
Why it is redundant?
It doesn't break any coding style rules here, but just separates loop.

Chocobo1 replied:
>but just separates loop.

IMO, for tiny code segment, the indention is enough for loop separation.

naikel replied:
Remember, everything I have done is because somewhere else in the same file is like that and it would look inconsistent.  The complete section is like this, only the second method is my code:
```
bool Session::hasPerTorrentRatioLimit() const
{
    foreach (TorrentHandle *const torrent, m_torrents)
        if (torrent->ratioLimit() >= 0) return true;

    return false;
}

bool Session::hasPerTorrentSeedingTimeLimit() const
{
    foreach (TorrentHandle *const torrent, m_torrents)
        if (torrent->seedingTimeLimit() >= 0) return true;

    return false;
}
```
And I think it looks better that way, like the original coder ;)

Chocobo1 replied:
>(only express our opinion).

This is exactly what I mean.

>Remember, everything I have done is because somewhere else in the same file is like that and it would look inconsistent. 

To be clear, I'm not forcing a change here.
Personal thought: I think the inconsistent of style isn't that critical anyway, unless someone rewrite the old code, it probably will stay there forever, so just focus on making the new code nice is enough...

Chocobo1 commented on src/gui/optionsdlg.cpp:
```diff
@@ -985,8 +990,19 @@ void OptionsDialog::loadOptions()
+        m_ui->checkMaxSeedingMinutes->setChecked(false);
+        m_ui->spinMaxSeedingMinutes->setEnabled(false);
+    }
+    m_ui->comboRatioLimitAct->setEnabled(session->globalMaxSeedingMinutes() >= 0 || session->globalMaxRatio() >= 0.);
```
please add parentheses

Chocobo1 commented on src/gui/transferlistwidget.cpp:
```diff
@@ -487,16 +487,24 @@ void TransferListWidget::setMaxRatioSelectedTorrents()
         torrent->setRatioLimit(ratio);
+
+        int seedingTime = (dlg.useDefault() ? BitTorrent::TorrentHandle::USE_GLOBAL_SEEDING_TIME: dlg.seedingTime());
+        torrent->setSeedingTimeLimit(seedingTime);
```
space between `USE_GLOBAL_SEEDING_TIME` and colon.

Chocobo1 commented on src/gui/updownratiodlg.cpp:
```diff
@@ -43,34 +46,78 @@ UpDownRatioDlg::UpDownRatioDlg(bool useDefault, qreal initialValue,
+void UpDownRatioDlg::accept()
+{
+    if (ui->torrentLimitButton->isChecked() && !ui->checkMaxRatio->isChecked() && !ui->checkMaxTime->isChecked())
+        QMessageBox::critical(0, tr("No share limit method selected"),
```
try find a parent to fill in, don't leave it 0

glassez replied:
**this** can be used as parent, isn't it?

evsh commented:
Oops, I've made a conflict. @naikel, could you rebase it one more (the last one) time, please?

glassez commented:
@evsh, I think we should no longer delay it, and merge it immediately after the conflicts will be resolved.

evsh commented:
Sure. I thought we agreed on this already. Came here to merge it and discovered the conflict.

naikel commented:
I fixed the last conflict @evsh @glassez @LordNyriox 

Now I'm curious why one of those last commits in master removed all classic Qt SIGNAL() and SLOT() and replaced them with that weird stuff...

LordNyriox commented:
@naikel, @evsh:  

>Now I'm curious why one of those last commits in master removed all classic Qt SIGNAL() and SLOT() and replaced them

I am not 100% clear on why this was done.  

If you read the comments on [<https://github.com/qbittorrent/qBittorrent/pull/6795>], there is apparently a "new syntax" for the `SIGNAL()` and `SLOT()` set.  

Also, I noticed the name used for the PR branch was `fix-file-widget`.  I have no idea what the change has to do with a "file widget", but there it is.

glassez commented:
>replaced them with that weird stuff

That weird stuff is pure c++.
This new Qt5 connection syntax has some advantages (e.g. compile time checking).

glassez commented:
@naikel, thank you.

thalieht commented:
Finally! Although i have no interest in this feature, thanks guys and @naikel for your patience and for not abandoning it. It was frustrating watching you doing rebase after rebase for so long and not being able to help.

Nivius commented:
finnaly!

when does it go live then? next update? :D

LordNyriox commented:
@Nivius:  It will not go live until qBittorrent v3.4.0.  

If there is a v3.3.13 release, this feature will NOT be included in it.

jeff15110168 commented:
omg been waiting for this feature forever. Please please please hoping that this rolls out soon :)

pret0rian commented:
Why this is still not in 3.4.0 alpha version? How many years you need to add so simple feature?

glassez commented:
>Why this is still not in 3.4.0 alpha version? How many years you need to add so simple feature?

This PR was merged onto master 2 months ago!

pret0rian commented:
> This PR was merged onto master 2 months ago!
So what? Im currently using 3.3.14 and still don't have that option...

glassez commented:
>>>Why this is still not in 3.4.0 alpha version? How many years you need to add so simple feature?
>>
>>This PR was merged onto master 2 months ago!
>
>So what? Im currently using 3.3.14 and still don't have that option...

Please decide which version you are asking: 3.4.0 alpha or 3.3.14!

pret0rian commented:
> Please decide which version you are asking: 3.4.0 alpha or 3.3.14!

Doesn't matter what version I use, you said it'3 merger 2 moth ago and I just say that newest version still does'nt have that, just check this out (no need to download 3.4.0 alpha):
https://github.com/qbittorrent/qBittorrent/compare/v3_3_x...35c51ad3b1765744d0c724331db60514123aabd5

jeff15110168 commented:
Would be great to have this in the 3.3.14 or 3.3.15 build :)

On Jul 22, 2017, at 5:05 AM, pret0rian <notifications@github.com<mailto:notifications@github.com>> wrote:


Please decide which version you are asking: 3.4.0 alpha or 3.3.14!

Doesn't matter what version I use, you said it'3 merger 2 moth ago and I just say that newest version still does'nt have that, just check this out (no need to download 3.4.0 alpha):
v3_3_x...35c51ad<https://github.com/qbittorrent/qBittorrent/compare/v3_3_x...35c51ad3b1765744d0c724331db60514123aabd5>

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/qbittorrent/qBittorrent/pull/4266#issuecomment-317168990>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AJbzgo3BNTf0JVc2xqYmrUMcbnZWlTsmks5sQbtogaJpZM4Gvh1I>.

zywo commented:
It was merged to master branch (currently 3.4.0alpha), it's not included in v3_3_x branch.

maca134 commented:
So how long until the general users can expect this feature?

LordNyriox commented:
@maca134:  

>So how long until the general users can expect this feature?  

Depending on how long the public beta for the release will operate, anywhere from two weeks to two months—rough estimate.

Nivius commented:
so, any estimates when 3.4 might drop?

corydorning commented:
not soon enough :(

jeff15110168 commented:
is this coming to 4.0?

corydorning commented:
Checking in again... This is the only thing keeping me from switching over from deluge which I'm not a fan of. But it has this feature.

Is it live yet?

corydorning commented:
So it looks like this is live but is it possible to do it based on category or tracker? Different trackers have different rules.
